app/models/user.rb

```diff
@@ -56,4 +56,14 @@ class User < ApplicationRecord
   def matches
     first_matches << second_matches
   end
+
+  def self.from_facebook(user_profile)
+    where(provider: 'facebook', uid: user_profile[:id]).first_or_create do |user|
+      user.password = Devise.friendly_token[0, 20]
+      user.uid = user_profile[:id]
```

Comment:
Since `user_profile` is coming from `FacebookService` you know which are the params. So you can make this method simpler using [assign_attributes](https://apidock.com/rails/v4.2.7/ActiveRecord/AttributeAssignment/assign_attributes):
```
def self.from_facebook(user_profile)
    where(provider: 'facebook', uid: user_profile[:id]).first_or_create do |user|
      user.password = Devise.friendly_token[0, 20]
      user.assign_attributes user_params.except('id')
    end
end

```